### PR TITLE
[TE] Support schedulable TIR compute definitions in TOPI

### PR DESCRIPTION
**Pull request by csullivan**
This PR adds `te.extern_primfunc` which provides the interface around TE ExternOp that allows a TVMScript defined schedulable TIR PrimFunc to be inlined into a TE compute graph. The result is that TIR can be used for compute definitions in Relay OpStrategies and, paired with meta-scheduler support in relay as introduced in #10578, these compute definitions can be scheduled and tuned as demonstrated in the attached tests.  

Prior to this, compute definitions were limited to those definable in TE only. As a consequence of this patch and ongoing improvements to TVMScript meta-programming (#11097), TOPI can be extended to include compute and scheduling functions targeting schedulable TIR uniformly.

### CODE REVIEW

**Comment by masahi** [comment]
cc @Hzfengsy

File: src/arith/domain_touched.cc [hunk 1]
```diff
@@ -34,18 +35,48 @@ namespace arith {
 
 using namespace tir;
 
+struct LoadAccess {
```
Comment (Hzfengsy) [inline src/arith/domain_touched.cc:38]: Can we use `using LoadAccess = std::vector<std::vector<IntSet>> set;`?

File: src/te/operation/create_primfunc.cc [hunk 2]
```diff
@@ -395,62 +395,75 @@ Stmt GenerateStmtFromExternOp(const te::ExternOp& extern_op, CreateFuncInfo* inf
...
-  CreateFuncInfo info(arg_list);
-  // Root body stmts.
-  Array<Stmt> root_stmts;
-  // Analyzer
-  arith::Analyzer analyzer;
+void InitializeBufferBinds(const Array<te::Operation>& ordered_ops, CreateFuncInfo* info) {
+  // Process any TE operations which contain user defined buffers
+  for (const auto& op : ordered_ops) {
+    // Initialize the tensor2buffer binds map with buffers defined by the te.extern
+    if (const auto extern_op = op.as<te::ExternOpNode>()) {
```
Comment (Hzfengsy) [inline src/te/operation/create_primfunc.cc:419]:
```suggestion
    if (const auto* extern_op = op.as<te::ExternOpNode>()) {
```

File: tests/python/unittest/test_meta_schedule_relay_tir_compute.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    @T.prim_func
+    def conv2d(
+        a: T.handle,
+        filt: T.handle,
+        b: T.handle,
+    ) -> None:
+        T.func_attr({"global_symbol": "main", "tir.noalias": True})
+        A = T.match_buffer(a, data_shape, dtype=dtype)
+        Filter = T.match_buffer(filt, weight_shape, dtype=dtype)
+        B = T.match_buffer(b, output_shape, dtype=dtype)
```
Comment (Hzfengsy) [inline tests/python/unittest/test_meta_schedule_relay_tir_compute.py:57]: It looks wired after `black` formatting. You can add `# fmt: off` and `# fmt: on` to turn off formatting for a part of python codes

**Review by Hzfengsy — APPROVED** [review]
Overall LGTM. Thanks @csullivan for such great contribution!!!

File: src/arith/domain_touched.cc (hunk 1 above)
Comment (csullivan) [reply src/arith/domain_touched.cc:42]: I'm currently using the struct type as sugar for compile-time lookup with std::get to index into the std::tuple. That said I took your advice an added an alias for the buffer touches. Happy to do it whichever way is preferable, I just found the tuple to be nice here.

**Review by csullivan — COMMENTED** [review]
Many thanks @Hzfengsy! PTAL again when you have the cycles.

File: tests/python/unittest/test_meta_schedule_relay_tir_compute.py (hunk 3 above)
Comment (csullivan) [reply tests/python/unittest/test_meta_schedule_relay_tir_compute.py:54]: Thanks for the suggestion, and agreed that it looked a bit weird. I've made the change.

File: src/arith/domain_touched.cc (hunk 1 above)
Comment (Hzfengsy) [reply src/arith/domain_touched.cc:42]: I don't have strong opinions about this. Both approaches are clear enough!

File: tests/python/unittest/test_tir_te_extern_primfunc.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    C: T.Buffer[(1,), "float32"],
+    A: T.Buffer[(16,), "float32"],
+    F: T.Buffer[(16,), "float32"],
+    D: T.Buffer[(2,), "float32"],
+    E: T.Buffer[(16,), "float32"],
+):
+    for i in T.serial(
+        0,
+        16,
+    ):
```
Comment (Hzfengsy) [inline tests/python/unittest/test_tir_te_extern_primfunc.py:145]:
```suggestion
    for i in range(16):
```

File: tests/python/unittest/test_tir_te_extern_primfunc.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    C: T.Buffer[(1,), "float32"],
+    A: T.Buffer[(16,), "float32"],
+    D: T.Buffer[(2,), "float32"],
+    E: T.Buffer[(16,), "float32"],
+    F: T.Buffer[(16,), "float32"],
+):
+    for i in T.serial(
+        0,
+        16,
+    ):
```
Comment (Hzfengsy) [inline tests/python/unittest/test_tir_te_extern_primfunc.py:103]:
```suggestion
    for i in range(16):
```

File: tests/python/unittest/test_tir_te_extern_primfunc.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+
+@T.prim_func
+def func_2(
+    C: T.Buffer[(1,), "float32"], A: T.Buffer[(16,), "float32"], D: T.Buffer[(2,), "float32"]
+):
+    for i in T.serial(
+        0,
+        16,
+    ):
```
Comment (Hzfengsy) [inline tests/python/unittest/test_tir_te_extern_primfunc.py:69]:
```suggestion
    for i in range(16):
```

File: tests/python/unittest/test_tir_te_extern_primfunc.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+# - PrimFunc with 1 arg, inplace update
+# - PrimFunc with buffer that uses custom storage_scope
+
+
+@T.prim_func
+def func_1(A: T.Buffer[(16,), "float32"], C: T.Buffer[(1,), "float32"]):
+    for i in T.serial(
+        0,
+        16,
+    ):
```
Comment (Hzfengsy) [inline tests/python/unittest/test_tir_te_extern_primfunc.py:40]:
```suggestion
    for i in range(16):
```

**Comment by junrushao1994** [comment]
@Hzfengsy for context, as a rule of thumb, usually we don't use "requested changes" if the comment is not emotional :-)  in our case, it's just some syntactic nitpicking, so let's just use "comment" as review type

**Comment by Hzfengsy** [comment]
Thanks for the reminder of @junrushao1994. And I'm sorry for bringing ambiguity to @csullivan. I will change the `Changes requested` into `Approve` and we can merge it if we want.

**Review by Hzfengsy — APPROVED** [review]

**Comment by csullivan** [comment]
Thanks @Hzfengsy. I think I prefer `T.serial` personally but I continue to be impressed by the TVMScript parser. I agree the linter made this a bit ugly. I want to add some additional tests in a follow up so I'll happily address the formatting then and consider the change to `range`.

**Pull request merged**